app/models/workflow.rb

```diff
@@ -1,34 +1,74 @@
 class Workflow
 
-  attr_accessor :steps
+  attr_accessor :first_step
 
   def self.build(definition)
     workflow = Workflow.new
-    workflow.steps = definition[:steps]
+
+    definition[:steps]
+      .each_with_index { |x, i| x[:next_steps] = [definition[:steps][i+1]].reject { |x| x.nil? } }
+      .each_with_index { |x, i| x[:method] = lambda { |e| Workflow.build_event_handler_for(x, e).receive e } }
+
+    workflow.first_step = definition[:steps].first
+    workflow.first_step[:method] = lambda { |e| Workflow.build_event_handler_for(workflow.first_step, e).fire e }
+
     workflow
   end
 
+  def self.build_given_a_hierarchy(definition)
+    workflow = Workflow.new
+
+    workflow.first_step = definition[:first_step]
+
+    set_up_the_method(workflow.first_step) { |e| Workflow.build_event_handler_for(workflow.first_step, e).fire e }
+
+    workflow
+  end
+
+  def self.set_up_the_method(step, &block)
+    step[:method] = block || lambda { |e| Workflow.build_event_handler_for(step, e).receive e }
+    step[:config] = {} if step[:config].nil?
+    return if step[:next_steps].nil?
+    step[:next_steps].each { |x| set_up_the_method(x) }
+  end
+
+  def self.build_event_handler_for(step, event)
```

Comment:
```suggestion
   def self.build_event_handler_for(step)
```

dont need to pass in `event`